File: 3.x.x/crates-utopia/emails/src/canonicals/fastmail.rs
use crate::canonicals::provider::{to_lower_case, Canonical, Provider};
use crate::EmailError;

/// Fastmail canonicalization (PHP `Fastmail`) - case + domain only.
#[derive(Debug, Clone, Copy, Default)]
pub struct Fastmail;

const SUPPORTED_DOMAINS: &[&str] = &["fastmail.com", "fastmail.fm"];
const CANONICAL_DOMAIN: &str = "fastmail.com";

impl Provider for Fastmail {
    fn supports(&self, domain: &str) -> bool {
        SUPPORTED_DOMAINS.contains(&domain)
    }

    fn get_canonical(&self, local: &str, _domain: &str) -> Result<Canonical, EmailError> {
        Ok(Canonical {
            local: to_lower_case(local),
            domain: CANONICAL_DOMAIN.to_string(),
        })
    }

    fn get_canonical_domain(&self) -> &'static str {
        CANONICAL_DOMAIN
    }

    fn get_supported_domains(&self) -> &'static [&'static str] {
        SUPPORTED_DOMAINS
    }
}

File: packages/emails/src/Canonicals/Providers/Fastmail.php
<?php

namespace Utopia\Emails\Canonicals\Providers;

use Utopia\Emails\Canonicals\Provider;

/**
 * Fastmail
 *
 * Handles Fastmail email normalization
 * - Preserves all characters in local part (no subaddress or dot removal)
 * - Normalizes to fastmail.com domain
 */
class Fastmail extends Provider
{
    private const SUPPORTED_DOMAINS = ['fastmail.com', 'fastmail.fm'];

    private const CANONICAL_DOMAIN = 'fastmail.com';

    public function supports(string $domain): bool
    {
        return in_array($domain, self::SUPPORTED_DOMAINS, true);
    }

    public function getCanonical(string $local, string $domain): array
    {
        // Convert to lowercase
        $normalizedLocal = $this->toLowerCase($local);

        // Fastmail doesn't remove subaddresses or dots
        // Just normalize case and domain

        return [
            'local' => $normalizedLocal,
            'domain' => self::CANONICAL_DOMAIN,
        ];
    }

    public function getCanonicalDomain(): string
    {
        return self::CANONICAL_DOMAIN;
    }

    public function getSupportedDomains(): array
    {
        return self::SUPPORTED_DOMAINS;
    }
}
